###### tags: `target_capture_workshop`

# SESSION 5: Species Trees

---

[TOC]

---

## 5.1 Why do we need a species tree?
Reconstructing species tree is an essential step into assessing relationships among our target species. When we generated trees from our targeted genes, we may notice that each gene tree only represents its own evolutionary history which differs from expected species tree (because of gene duplication, lost, convention, historical hybridization, incomplete lineage sorting, and presence of deep coalescence). In other word, these gene trees may differ from each other and not agree with species tree which reflects the real phylogeny of associated species. In order to acquire a consensus phylogeny to study in related species, we would like to infer species tree using specific methods.

![Figure1DeepCoalescence](https://github.com/gudusanjiao/targetseqphylo/blob/main/miscellaneous/pic2.png?raw=true "Fig 1 Deep Coalescence")

Figure 1. Deep coalescence may cause discordance between gene tree and species tree.

More information about why gene trees often do not agree with species tree, you can refer to [Szöllősi et al.](https://doi.org/10.1093/sysbio/syu048)
## 5.2 How to infer phylogenies of species?
There are two major ways in inferring species tree. First and simple one is to use a concatenated dataset, which refers to **concatenation**. To apply this method, every gene will be added together and aligned into a supermatrix. Then, further phylogenetic inferences will be deployed onto this concatenated gene which will be treated as a single gene complex. 

However, evidence implied that the concatenation is not always correctly reconstruct the phylogeny. It lacks modeling deep coalescence which creates disagreement with the true phylogeny. This is mainly caused by a majority of genes may not share the same history as species do. Thus, when using concatenation method, signals from discordant genes would outweighing the less coalescence ones.

In order to solve the problem of deep coalescence in species tree reconstruction, several methods have been developed to either **model the process of deep coalescence** or taking **shortcuts** where deep coalescence trees still consist with real species tree. Full modeling of deep coalescence is more accurate but cannot take large datasets due to the complexity of computational time. Since genomic level sequencing is more and more prevalent nowadays (including our target sequencing strategy), shortcuts methods are becoming popular because they are usually effective and relatively reliable.

![image](https://user-images.githubusercontent.com/16470742/143940257-bec4517c-f3df-4b03-8e38-fc6239745896.png "Fig 2 Species Tree Methods")

Figure 2. Schematic of the concatenation and coalescent paradigms in phylogenetics. ([Liu et al.](https://nyaspubs.onlinelibrary.wiley.com/doi/full/10.1111/nyas.12747))

## 5.3 Practicing species tree method using ASTRAL
In today's workshop, we will use [ASTRAL](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-018-2129-y) first, one of the shortcuts methods, to infer our species tree from our gene tree datasets. 

ASTRAL is a Java phylogenetics program that uses the frequency of quartets in gene trees to estimate a species tree. It offers a one command line solution under Linux. Comparing these quartets is not computational intense so this program may finish in a short time. However, to be consistent to other sessions in our workshop, we will still run this under HPCC job submitting system.

Before running on HPCC, don't forget to request 4 CPUs for today's work. Remember, we cannot run jobs on login node.
```bash
interactive -p nocona -c 4
```

For ASTRAL, we do not need to have any containers since it is an independent Java application. Any pre-made ASTRAL version is good to be applied for this job. Here, we use the ASTRAL that installed in my home directory. If you are interested in getting an copy under your folder, you may refer to their [github](https://github.com/smirarab/ASTRAL). According to their description, we only need to download the zip file and unzip it.

The input file is a merged Newick format tree file under `/home/nhu/TC_workshop/input_gene_tree.tre`. This file is generated by merging all iqtree output `.treefile` into one single `.tre` file. You may check the structure of this file by using `head [file name]`. You may see that each line of this file is a gene tree in Newick format.

```bash
java -jar /home/nhu/software/astral/ASTRAL/astral.5.7.8.jar -i [unrooted gene tree file as input, .tre] -o [output species tree file name, .tre]
```

This program should be finished within one minute. Output file will also be in a Newick tree-form text. You may download it to local for further viewing and editing.
### Action 5.3.1
*Use ASTRAL to generate results from gene trees*
```

```

## 5.4 Practicing species tree method using SVDQuartets (on PAUP)
SVDQuartets is also a method assessing quartets trees in input file. Instead of using gene trees, it takes single nucleotides onto quartets to calculate SVD (Singular Value Decomposition). Thus, the input file is a concatenated FASTA (actually should be coverted into NEXUS format, an universal phylogeny prep format for a lot of programs). To clarify, the input is concatenated but this is not a concatenation method.

![image](https://github.com/gudusanjiao/targetseqphylo/blob/main/miscellaneous/svdq.png?raw=true)

Figure 3. Quartets and SVDQuartets. ([Source](https://www.asc.ohio-state.edu/kubatko.2//SVDQ_Intro.pdf))

PAUP (Phylogenetic Analysis Using PAUP) is a GUI software integrating different types of phylogenetic analysis, including SVDQuartets. Computers in this room may not be able to install a software. If we cannot practice during our workshop, we have finished tree files for you to compare. You may try it on your own machine.

> If we cannot install PAUP on public machine, pre-made trees are available under `/home/nhu/TC_workshop/all_concat.tre` and `all_concat_SVDQ.tre`.

### Action 5.4.1
*Download PAUP from [PAUP](http://phylosolutions.com/paup-test/). Choose the Windows GUI version. Install it.*

### Action 5.4.2
*Open NEXUS file `all_concat.nex` in PAUP (download through FTP in HPCC `/home/nhu/TC_workshop/all_concat.nex`). Wait for the program to excute the file. When finished, you may see the information of input file as well as 'Completed' in Display window.*

> To save time, we can remove some taxa, leave 20 taxa for tree construction. This will save a huge amount of time especially for bootstrapping. In order to do this, select [Data - Delete/Restore Taxa], in the popped up window, move all the taxa below #20 to the right. By doing this, we only include 20 taxa in our analysis.

### Action 5.4.3
*Click on [Analysis - SVDQuartets]. In popped up window, for tree model option, choose 'shared tree for all sites'. This will generate a tree similar to using a **concatenation method** by assuming all sites evolved under the same tree. We also click the Bootstrapping check to generate bootstrapping values for our tree.*

*After program finishing running, click [Trees - Save Trees To File]. Choose Newick format to save.*

### Action 5.4.4
*Repeat analysis, but using 'Multispecies coalescent' methods. This will let program to assess coalescense using SVDQuartets. Still, save the output tree.*

## 5.5 Interpreting ASTRAL results
After generating species tree from ASTRAL and PAUP, we may view and edit it through some tree viewer software by rerooting, collapsing clades, re-coloring, enhancing branches, showing branch length etc. The goal of customizing it is to better convey our result to readers. [FigTree](https://github.com/rambaut/figtree/releases) is a Java-based tree viewer offering numbers of options in customize phylogenetic trees. If you want to install it on your own computer, you need a [Java](https://www.java.com/en/download/) environment installed first.

When you load Newick-form Tree file (.tre) in FigTree, you will be asked for the label of node/branch values. Notice that, support values in ASTRAL species tree means the **Local Posterior Probability**, which indicates how well the shown quartet is compared to the other two possible quartets at that branch. The branch lengths are **coalescent units** in related to ancestral population size and the amount of change found on the gene trees. 

### Action 5.5.1
*Download Figtree. Unzip it. And transfer ASTRAL output to local folder to view it with FigTree*

As we said above, FigTree offers tons of options to modify our initial input file. There are several buttons on the top of the window for quick functions to modify a tree. First action to try is to reroot our output species tree based on known outgroup. To apply this, you may select the outgroup branch to make it highlighted, then click "Reroot" button on the top. You may notice that the order of our species tree seems to be changed. Actually, the topology of this tree is not altered at all, which means the relationships between each species on the tree remain the same.

### Action 5.5.2
*Reroot species tree by selected outgroup*

After rerooting our tree, we also want to show the node statistics for each bifurcations. In FigTree, we can click on triangle beside "Node Labels" on sidebar to expand some options available for node information display. In order to display probability on node, we will choose our named label (software asked when you opened the tree file) on "Display". Then, we can customize the font, size, color and format etc. to better convey the information.

### Action 5.5.3
*Display node values in a nice-looking way*

For some phylogeny trees in papers, they often highlighted some branches to inform readers for specific reasons such as shared ancestry, co-evolution events, and homoplasy traits. These highlightings usually enhance the effectiveness in conveying information. To practise this, we can click on triangle beside "Tip Labels" on side bar. Then, we select one branch and click on color option on the side. Choose a color that you think will be appropriate for your tree.

### Action 5.5.4
*Highlight branches that belong to the same taxanomy groups*

There are a lot of other options for you to customize your phylogenetic trees in FigTree. We would like to leave this oppotunity of exploring it to you. You can try but not be limited to change the size of tip labels, change size of branches, change appearance of the tree into cladogram, make highlighted blocks, add scales to tree, and rename tips etc. The ultimate goal is to help yourself interpret the tree and help other better understand what you want to exhibit.

### Action 5.5.5
*Customize both your ASTRAL and SVDQuatets trees in FigTree and try to interpret and compare*

